ledger-service/http-json/src/main/scala/com/digitalasset/http/WebSocketService.scala

```diff
@@ -222,61 +231,42 @@ object WebSocketService {
 
     def acsRequest(
         maybePrefix: Option[domain.StartingOffset],
-        request: A,
-    ): Option[A]
+        resolvedRequest: R,
+    ): Option[R]
 
     /** Perform any necessary adjustment to the request based on the prefix
       */
     def adjustRequest(
         prefix: Option[domain.StartingOffset],
-        request: A,
-    ): A
+        resolvedRequest: R,
+    ): R
 
     /** Specify the offset from which the live part of the query should start
       */
     def liveStartingOffset(
         prefix: Option[domain.StartingOffset],
-        request: A,
+        resolvedRequest: R,
     ): Option[domain.StartingOffset]
-
   }
 
-  private sealed trait UnsupportedOrResolvedQueryTag
-  private type UnsupportedOrResolvedQuery =
-    (Unsupported \/ ResolvedQuery) @@ UnsupportedOrResolvedQueryTag
-  private val UnsupportedOrResolvedQuery = Tag.of[UnsupportedOrResolvedQueryTag]
-
-  private implicit val `Unsupported or ResolvedQuery semigroup`
-      : Semigroup[UnsupportedOrResolvedQuery] = {
-    import ResolvedQuery._
-    UnsupportedOrResolvedQuery subst Semigroup.instance {
-      case (-\/(CannotQueryBothTemplateIdsAndInterfaceIds), _) |
-          (_, -\/(CannotQueryBothTemplateIdsAndInterfaceIds)) =>
-        -\/(CannotQueryBothTemplateIdsAndInterfaceIds)
-
-      case (-\/(CannotQueryManyInterfaceIds), _) | (_, -\/(CannotQueryManyInterfaceIds)) =>
-        -\/(CannotQueryManyInterfaceIds)
-
-      case (-\/(CannotBeEmpty), _) | (_, -\/(CannotBeEmpty)) => -\/(CannotBeEmpty)
-
-      case (\/-(ByInterfaceId(_)), \/-(ByTemplateId(_)) | \/-(ByTemplateIds(_))) |
-          (\/-(ByTemplateId(_)) | \/-(ByTemplateIds(_)), \/-(ByInterfaceId(_))) =>
-        -\/(CannotQueryBothTemplateIdsAndInterfaceIds)
+  final case class ResolvedSearchForeverRequest(
+      resolvedQuery: ResolvedQuery,
```

Comment:
Normally, the nature of this `resolvedQuery`'s relationship to those contained in the component `ResolvedSearchForeverQuery`s would be defined by the associated semigroup. That isn't the case here, so is there an invariant regarding that relationship that you can describe here?

Reply:
All `ResolvedSearchForeverQuery`s must be resolved. The `resolvedQuery` of `ResolvedSearchForeverRequest` is all the resolved sets of all the `ResolvedSearchForeverQuery`s combined.